master/internal/api_workspace.go

```diff
@@ -521,6 +554,117 @@ func (a *apiServer) PatchWorkspace(
+		CanAddWorkspaceNamespaceBindings(ctx, *curUser, int32(wkspID)); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	err = a.validateRequestClusterName(req.ClusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify that the namespace exists in the Kubernetes cluster for the corresponding RM.
+	err = a.m.rm.VerifyNamespaceExists(*req.NamespaceName, req.ClusterName)
+	if err != nil {
+		return nil, fmt.Errorf("error verifying Kubernetes namespace: %w", err)
+	}
+
+	if req.ClusterName == nil {
+		emptyClusterName := ""
+		req.ClusterName = &emptyClusterName
+	}
+
+	// Change the workspace-namespace binding for the given cluster if it already exists. If no
+	// namespace is bound to the workspace for the specified cluster, add one.
+	findWsNsQuery := tx.NewSelect().Model(&model.WorkspaceNamespace{}).
+		Where("workspace_id = ?", wkspID).
+		Where("cluster_name = ?", req.ClusterName)
+
+	var wsns model.WorkspaceNamespace
+	err = findWsNsQuery.Scan(ctx, &wsns)
+	if err != nil {
+		if err != sql.ErrNoRows {
+			return nil, fmt.Errorf("error getting the current workspace-namespace binding: %w", err)
+		}
+		// The workspace has no namespace binding for the specified cluster, so we add one.
+		wsns = model.WorkspaceNamespace{
+			NamespaceName: *req.NamespaceName,
+			ClusterName:   *req.ClusterName,
+			WorkspaceID:   wkspID,
+		}
+		err = tx.NewInsert().Model(&wsns).Scan(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("error adding workspace-namespace binding to database: %w", err)
+		}
+	} else {
+		// Update the existing workspace-namespace binding for the specified cluster.
+		_, err = tx.NewUpdate().Model(&model.WorkspaceNamespace{}).
+			Set("namespace_name = ?", *req.NamespaceName).
+			Where("workspace_id = ?", wkspID).
+			Where("namespace_name = ?", *req.NamespaceName).
+			Where("cluster_name = ?", req.ClusterName).
+			Exec(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("could not update workspace-namespace binding: %w", err)
+		}
+	}
+	err = findWsNsQuery.Scan(ctx, &wsns)
+	if err != nil {
+		return nil, err
+	}
+	wsnsProto := wsns.ToProto()
+	res := apiv1.AddWorkspaceNamespaceBindingResponse{WorkspaceNamespace: wsnsProto}
+
+	return &res, nil
+}
+
+func (a *apiServer) AddWorkspaceNamespaceBinding(ctx context.Context,
+	req *apiv1.AddWorkspaceNamespaceBindingRequest,
+) (*apiv1.AddWorkspaceNamespaceBindingResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := a.GetWorkspaceByID(ctx, req.Id, *curUser, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if req.ClusterName != nil && req.NamespaceName == nil {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"Must specify either an existing Kubernetes namespace")
+	}
+
+	var finalRes *apiv1.AddWorkspaceNamespaceBindingResponse
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		res, err := a.addWorkspaceNamespaceBinding(ctx, req, &tx, int(w.Id))
```

Comment:
I don't think it makes sense to have a helper function when this function isn't too big. We are calling GetUser in both functions and also in my PR I am making pluralizing this endpoint, so I would have to iterate over the Map twice

Reply:
Imo the helper function is useful because we call `addWorkspaceNamespaceBinding` (now changed to `setWorkspaceNamespaceBinding`) in both the `PostWorkspace` request and the `AddWorkspaceNamespaceBinding`/ `SetWorkspaceNamespaceBinding` request, and I don't really want to make a direct call to the `AddWorkspaceNamespaceBinding` handler within the `PostWorkspace` handler.
I think when there's identical logic that needs to happen in two different service endpoints, we should abstract the logic and call it separately with a helper function.